homeassistant/components/binary_sensor/xknx.py

```diff
@@ -0,0 +1,164 @@
+"""
+Support for KNX/IP binary sensors via XKNX.
+
+For more details about this platform, please refer to the documentation at
+https://home-assistant.io/components/binary_sensor.xknx/
+"""
+import asyncio
+import voluptuous as vol
+
+from homeassistant.components.xknx import DATA_XKNX
+from homeassistant.components.binary_sensor import PLATFORM_SCHEMA, \
+    BinarySensorDevice
+from homeassistant.const import CONF_NAME
+import homeassistant.helpers.config_validation as cv
+from homeassistant.helpers.script import Script
+
+CONF_ADDRESS = 'address'
+CONF_DEVICE_CLASS = 'device_class'
+CONF_SIGNIFICANT_BIT = 'significant_bit'
+CONF_DEFAULT_SIGNIFICANT_BIT = 1
+CONF_AUTOMATION = 'automation'
+CONF_HOOK = 'hook'
+CONF_DEFAULT_HOOK = 'on'
+CONF_COUNTER = 'counter'
+CONF_DEFAULT_COUNTER = 1
+CONF_ACTION = 'action'
+
+CONF__ACTION = 'turn_off_action'
+
+DEFAULT_NAME = 'XKNX Binary Sensor'
+DEPENDENCIES = ['xknx']
+
+AUTOMATION_SCHEMA = vol.Schema({
+    vol.Optional(CONF_HOOK, default=CONF_DEFAULT_HOOK): cv.string,
+    vol.Optional(CONF_COUNTER, default=CONF_DEFAULT_COUNTER): cv.port,
+    vol.Required(CONF_ACTION, default=None): cv.SCRIPT_SCHEMA
+})
+
+AUTOMATIONS_SCHEMA = vol.All(
+    cv.ensure_list,
+    [AUTOMATION_SCHEMA]
+)
+
+PLATFORM_SCHEMA = PLATFORM_SCHEMA.extend({
+    vol.Required(CONF_ADDRESS): cv.string,
+    vol.Optional(CONF_NAME, default=DEFAULT_NAME): cv.string,
+    vol.Optional(CONF_DEVICE_CLASS): cv.string,
+    vol.Optional(CONF_SIGNIFICANT_BIT, default=CONF_DEFAULT_SIGNIFICANT_BIT):
+        cv.positive_int,
+    vol.Optional(CONF_AUTOMATION, default=None): AUTOMATIONS_SCHEMA,
+})
+
+
+@asyncio.coroutine
+def async_setup_platform(hass, config, add_devices,
+                         discovery_info=None):
+    """Set up binary sensor(s) for XKNX platform."""
+    if DATA_XKNX not in hass.data \
+            or not hass.data[DATA_XKNX].initialized:
+        return False
+
+    if discovery_info is not None:
+        yield from add_devices_from_component(hass, add_devices)
+    else:
+        yield from add_devices_from_platform(hass, config, add_devices)
+
+    return True
+
+
+@asyncio.coroutine
+def add_devices_from_component(hass, add_devices):
```

Comment:
Need not be a coroutine. Only if you Need `yield from` inside a function, that need to be a coroutine. If a function only run inside loop add `@callback`. Add also `async_` sufix to function like this.